lib/services/scraper/scraper.go

```diff
@@ -0,0 +1,148 @@
+package scraper
+
+import (
+	"fmt"
+	"net/url"
+
+	"go-google-scraper-challenge/helpers/log"
+	"go-google-scraper-challenge/lib/models"
+
+	"github.com/gocolly/colly"
+)
+
+type Scraper struct {
+	Result *models.Result
+}
+
+var selectors = map[string]string{
+	"wholePage":    "html",
+	"topImageAds":  "#tvcap .plantl a.pla-unit-title-link",
+	"topLinkAds":   "#tvcap .d5oMvf > a",
+	"sideImageAds": ".cu-container a.plantl.clickable-card",
+	"nonAds":       "#search .yuRUbf > a",
+	"bottomAds":    "#bottomads .uEierd > a",
+}
+
+const GoogleSearchUrl = "http://www.google.com/search?q=%s"
+
+func (s *Scraper) Run() error {
+	escapedKeyword := url.QueryEscape(s.Result.Keyword)
+	url := fmt.Sprintf(GoogleSearchUrl, escapedKeyword)
+
+	return s.startScraping(url)
+}
+
+func (s *Scraper) startScraping(url string) error {
+	collector := colly.NewCollector()
+
+	collector.OnRequest(s.requestHandler)
+	collector.OnResponse(s.responseHandler)
+	collector.OnError(s.errorHandler)
+
+	collector.OnHTML(selectors["wholePage"], s.wholePageCollector)
+
+	collector.OnHTML(selectors["nonAds"], func(e *colly.HTMLElement) {
+		s.addNonAdLinkToResult(e)
+	})
+
+	collector.OnHTML(selectors["topImageAds"], func(e *colly.HTMLElement) {
+		s.addAdLinkToResult(models.AdLinkTypeImage, models.AdLinkPositionTop, e)
+	})
+
+	collector.OnHTML(selectors["topLinkAds"], func(e *colly.HTMLElement) {
+		s.addAdLinkToResult(models.AdLinkTypeLink, models.AdLinkPositionTop, e)
+	})
+
+	collector.OnHTML(selectors["sideImageAds"], func(e *colly.HTMLElement) {
+		s.addAdLinkToResult(models.AdLinkTypeImage, models.AdLinkPositionSide, e)
+	})
+
+	collector.OnHTML(selectors["bottomAds"], func(e *colly.HTMLElement) {
+		s.addAdLinkToResult(models.AdLinkTypeLink, models.AdLinkPositionBottom, e)
+	})
```

Comment:
This part looks straightforward to me 😉
But this part is not including with the part about the rollback methodology 🥲
In the scenario that the service is running and it is crashed somehow ... all the previously added records are still there. ☠️

The way to improve this part is quite big ... here is my idea 👨‍💻

1. Replace all the logic about creating scraped data to the database into a bunch of variables.
2. Create another service for dealing with data creation.
3. Assign those data into its table with the association and save at once or create those in table by table wrapping with the transaction.

Reply:
Thanks updated in [7097da1](https://github.com/carryall/go-google-scraper-challenge/pull/104/commits/7097da12c8b8a193aad974282f954c436ce045cb)